tests/integration/ha_tests/helpers.py

```diff
@@ -498,32 +508,118 @@ async def mongod_ready(ops_test, unit_ip) -> bool:
+        log = client.admin.command("getLog", "global")
+        client.close()
 
-        if "msg" not in item:
+        if unit.public_address == old_primary:
             continue
 
-        # this message indicates that the previous primary performed a repl step down operation.
-        # its important to check that this step down was performed after the sigterm opteration
-        # was performed, as it could have been performed at an earlier time for another reason.
-        step_down_time = convert_time(item["t"]["$date"])
-        if (
-            item["msg"] == "Starting an election due to step up request"
-            and step_down_time > sigterm_time
-        ):
-            return True
+        for item in log["log"]:
+            item = json.loads(item)
+
+            if "msg" not in item:
+                continue
+
+            # this message indicates that the previous primary performed a repl step down
+            # operation. its important to check that this step down was performed after the
+            # sigterm opteration was performed, as it could have been performed at an earlier
+            # time for another reason.
+            step_down_time = convert_time(item["t"]["$date"])
+            if (
+                item["msg"] == "Starting an election due to step up request"
+                and step_down_time >= sigterm_time
+            ):
+                return True
 
     return False
 
 
+async def all_db_processes_down(ops_test: OpsTest) -> bool:
+    """Verifies that all units of the cgarm do not have the DB process running."""
+    app = await app_name(ops_test)
+
+    try:
+        for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+            with attempt:
+                for unit in ops_test.model.applications[app].units:
+                    search_db_process = f"run --unit {unit.name} ps aux | grep {DB_PROCESS}"
+                    _, processes, _ = await ops_test.juju(*search_db_process.split())
+
+                    # `ps aux | grep {DB_PROCESS}` is a process on it's own and will be shown in
+                    # the output of ps aux, hence it it is important that we check if there is
+                    # more than one process containing the name `DB_PROCESS`
+                    # splitting processes by "\n" results in an empty line, hence we need to
+                    # remove one to account for the extra entry
+                    if len(processes.split("\n")) - 1 > 1:
+                        raise ProcessRunningError
+    except RetryError:
+        return False
+
+    return True
+
+
+async def update_restart_delay(ops_test: OpsTest, unit, delay: int):
+    """Updates the restart delay in the DB service file.
+
+    When the DB service fails it will now wait for `dalay` number of seconds.
+    """
+    # load the service file from the unit and update it with the new delay
+    await unit.scp_from(source=MONGOD_SERVICE_DEFAULT_PATH, destination=TMP_SERVICE_PATH)
```

Comment:
If you choose to change the code to use `NamedTemporaryFile` (see my other comment) you can do something like below:
```python
with NamedTemporaryFile(mode="w") as f:
    await unit.scp_from(source=MONGOD_SERVICE_DEFAULT_PATH, destination=f.name)
    # Then run other steps like changing the file and uploading it again to the unit.
```

Reply:
wasn't quite able to get this to work so I will keep this as is for now